Configuration binding source generator fails with CS8598 error

### Description

Attempting to enable the Configuration binding source generator with a struct used as a custom configuration type causes the source generator to emit code that fails to compile with a CS8598 error.

```
Repro\Microsoft.Extensions.Configuration.Binder.SourceGeneration\Microsoft.Extensions.Configuration.Binder.SourceGeneration.ConfigurationBindingGenerator\GeneratedConfigurationBinder.g.cs(122,17): error CS8598: The suppression operator is not allowed in this context [Repro\Repro.csproj]
```

The method generated is below. The failing line is `element! = temp4;`.

```csharp
public static void BindCore(IConfiguration configuration, ref IDictionary<string, Geolocation> obj, BinderOptions? binderOptions)
{
    if (obj is null)
    {
        throw new ArgumentNullException(nameof(obj));
    }

    foreach (IConfigurationSection section in configuration.GetChildren())
    {
        if (!(obj.TryGetValue(section.Key!, out Geolocation element)))
        {
            element = InitializeGeolocation(section, binderOptions);;
        }
        var temp4 = InitializeGeolocation(configuration, binderOptions);;
        BindCore(section, ref temp4, binderOptions);
        element! = temp4; // CS8598 occurs here
        obj[section.Key!] = element;
    }
}
```


### Reproduction Steps

1. Clone [martincostello/config-binding-source-generator-cs8598-repro](https://github.com/martincostello/config-binding-source-generator-cs8598-repro)
1. Run `dotnet build`


### Expected behavior

The application compiles and configuration binding works equivalently to runtime binding.


### Actual behavior

The application fails to compile with a CS8598 error.


### Regression?

No.

### Known Workarounds

None.

### Configuration

.NET SDK 8.0.100-preview.6.23330.14

### Other information

_No response_

--- a/src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Emitter/CoreBindingHelper.cs
+++ b/src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Emitter/CoreBindingHelper.cs
@@ -731,7 +731,7 @@ namespace Microsoft.Extensions.Configuration.Binder.SourceGeneration
                                 """);
                         }
 
-                        EmitBindCoreCall(elementType, $"{Identifier.element}!", Identifier.section, InitializationKind.None);
+                        EmitBindCoreCall(elementType, $"{Identifier.element}", Identifier.section, InitializationKind.None);
                         _writer.WriteLine($"{objIdentifier}[{parsedKeyExpr}] = {Identifier.element};");
                     }
 

--- a/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.TestClasses.cs
+++ b/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.TestClasses.cs
@@ -3,6 +3,8 @@
 
 using System;
 using System.Collections.Generic;
+using System.ComponentModel;
+using System.Diagnostics.CodeAnalysis;
 using System.Globalization;
 using System.Linq;
 using Microsoft.Extensions.Configuration;
@@ -663,6 +665,60 @@ namespace Microsoft.Extensions
             public StructWithParameterlessAndParameterizedCtor(int myInt) => MyInt = 10;
 
             public int MyInt { get; }
+        }
+
+        [TypeConverter(typeof(GeolocationTypeConverter))]
+        public struct Geolocation : IEquatable<Geolocation>, IParsable<Geolocation>
+        {
+            public static readonly Geolocation Zero = new(0, 0);
+
+            public Geolocation(double latitude, double longitude)
+            {
+                Latitude = latitude;
+                Longitude = longitude;
+            }
+
+            public double Latitude { get; set; }
+
+            public double Longitude { get; set; }
+
+            private sealed class GeolocationTypeConverter : TypeConverter
+            {
+                public override bool CanConvertFrom(ITypeDescriptorContext? context, Type sourceType)
+                {
+                    if (sourceType == typeof(string) || sourceType == typeof(Geolocation))
+                    {
+                        return true;
+                    }
+
+                    return base.CanConvertFrom(context, sourceType);
+                }
+
+                public override object? ConvertFrom(ITypeDescriptorContext? context, CultureInfo? culture, object value)
+                {
+                    if (value is string s)
+                    {
+                        return Parse(s, culture);
+                    }
+                    else if (value is Geolocation geolocation)
+                    {
+                        return geolocation;
+                    }
+
+                    return base.ConvertFrom(context, culture, value);
+                }
+            }
+
+            public bool Equals(Geolocation other) => Latitude == other.Latitude && Longitude == other.Longitude;
+
+            public static Geolocation Parse(string s, IFormatProvider? provider) => throw new NotImplementedException();
+
+            public static bool TryParse([NotNullWhen(true)] string? s, IFormatProvider? provider, [MaybeNullWhen(false)] out Geolocation result) => throw new NotImplementedException();
+        }
+
+        public class GeolocationWrapper
+        {
+            public Geolocation Location { get; set; }
         }
     }
 }

--- a/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.cs
+++ b/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.cs
@@ -1900,5 +1900,70 @@ if (!System.Diagnostics.Debugger.IsAttached) { System.Diagnostics.Debugger.Launc
             GenericOptionsWithParamCtor<string> obj2 = configuration.Get<GenericOptionsWithParamCtor<string>>();
             Assert.Equal("MyString", obj2.Value);
         }
+
+        [Fact]
+        public void ObjWith_IParsableT_And_TypeConverter()
+        {
+            var configuration = TestHelpers.GetConfigurationFromJsonString("""
+                {
+                    "Location":
+                    {
+                        "Latitude": 3,
+                        "Longitude": 4,
+                    }
+                }
+                """);
+
+            // Neither IParsableT or TypeConverter impl are honored (https://github.com/dotnet/runtime/issues/83599).
+
+            GeolocationWrapper obj = configuration.Get<GeolocationWrapper>();
+            ValidateGeolocation(obj.Location);
+
+            configuration = TestHelpers.GetConfigurationFromJsonString(""" { "Geolocation": "3, 4", } """);
+            obj = configuration.Get<GeolocationWrapper>();
+            Assert.Equal(Geolocation.Zero, obj.Location);
+        }
+
+        [Fact]
+        public void ComplexObj_As_Dictionary_Element()
+        {
+            var configuration = TestHelpers.GetConfigurationFromJsonString("""
+                {
+                    "First":
+                    {
+                        "Latitude": 3,
+                        "Longitude": 4,
+                    }
+                }
+                """);
+
+            Geolocation obj = configuration.Get<IDictionary<string, Geolocation>>()["First"];
+            ValidateGeolocation(obj);
+
+            obj = configuration.Get<IReadOnlyDictionary<string, Geolocation>>()["First"];
+            ValidateGeolocation(obj);
+        }
+
+        [Fact]
+        public void ComplexObj_As_Enumerable_Element()
+        {
+            var  configuration = TestHelpers.GetConfigurationFromJsonString("""{ "Enumerable": [{ "Latitude": 3, "Longitude": 4 }] }""")
+                .GetSection("Enumerable");
+
+            Geolocation obj = configuration.Get<IList<Geolocation>>()[0];
+            ValidateGeolocation(obj);
+
+            obj = configuration.Get<Geolocation[]>()[0];
+            ValidateGeolocation(obj);
+
+            obj = configuration.Get<IReadOnlyList<Geolocation>>()[0];
+            ValidateGeolocation(obj);
+        }
+
+        private void ValidateGeolocation(Geolocation location)
+        {
+            Assert.Equal(3, location.Latitude);
+            Assert.Equal(4, location.Longitude);
+        }
     }
 }

